File: RemoveDuplicatesSortedArrayII.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

public class RemoveDuplicatesSortedArrayII {
    public int removeDuplicates(int[] nums) {
        if (nums.length == 0 || nums == null) {
            return 0;
        }
        int repCount = 1;
        int slowPtr = 1;
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] == nums[i - 1]) {
                repCount++;
            } else {
                repCount = 1;
            }
            if (repCount <= 2) {
                nums[slowPtr] = nums[i];
                slowPtr++;
            }
        }
        return slowPtr;
    }
}

File: RemoveDuplicatesSortedArrayII.cs
// Time Complexity :O(N)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :Yes
// Three line explanation of solution in plain english:

// This code removes duplicates from an array, allowing at most two occurrences of each element.

// It counts occurrences of each element, retaining them if they occur less than or equal to twice.

// The resulting array without duplicates is returned, and its length is determined.

public class Solution {
    public int RemoveDuplicates(int[] nums) {
        if(nums == null || nums.Length ==0) return 0;
        int n = nums.Length;
        int slow =0;
        int count =0;

        for(int i=0; i<n ; i++)
        {
            if(i == 0 || nums[i] == nums[i-1])
            {
                count++;
            }
            else
            {
                count =1;
            }
            if(count <=2)
            {
                nums[slow] = nums[i]; //O(1)
                slow++;
            }
        }
        return slow;

    }
}